openvdb/openvdb/points/PointRasterizeFrustum.h

```diff
@@ -0,0 +1,1805 @@
+    const math::Transform& firstTransform() const;
+    const math::Transform& lastTransform() const;
+
+    void setShutter(float start, float end);
+    float shutterStart() const;
+    float shutterEnd() const;
+
+private:
+    std::deque<math::Transform> mTransforms;
+    std::deque<float> mWeights;
+    // default to 180 degree film shutter
+    float mShutterStart = -0.25f;
+    float mShutterEnd = 0.25f;
+}; // class RasterCamera
+
+
+/// @brief A group of shared settings to be used in the Volume Rasterizer
+/// @param scaleByVoxelVolume   scale particle contributions by the volume of the receiving voxel
+/// @param velocityAttribute    the name of the velocity attribute
+/// @param velocityMotionBlur   bake the point velocities into the volume
+/// @param clipToFrustum        if enabled and the transform is a frustum transform, eliminate
+///                             points whose position does not lie within the frustum
+/// @param clipBBox             an optional world-space bounding box to clip the points
+///                             during rasterization
+/// @param clipMask             an optional mask, each point samples the mask using a
+///                             nearest-neighbor sampling and is only rasterized if active
+/// @param invertMask           if mask is provided, only rasterize if sample is inactive
+/// @param framesPerSecond      the global value for frames / second for computing motion blur
+/// @param threaded             if enabled, use threading to accelerate rasterization
+/// @note rasterization can clip can using any combination of bounding box, mask and frustum
+struct FrustumRasterizerSettings
+{
+    FrustumRasterizerSettings() = delete;
+
+    explicit FrustumRasterizerSettings(const math::Transform& _transform)
+        : transform(new math::Transform(_transform))
+        , camera(_transform) { }
+
+    math::Transform::Ptr transform;
+    RasterCamera camera;
+    float threshold = 1e-6f;
+    bool scaleByVoxelVolume = false;
+    Name velocityAttribute = "v";
+    bool useRadius = false;
+    float radiusScale = 1.0f;
+    Name radiusAttribute = "pscale";
+    bool accurateFrustumRadius = false;
+    bool accurateSphereMotionBlur = false;
+    bool velocityMotionBlur = false;
+    float framesPerSecond = 24.0f;
+    int motionSamples = 2;
+    bool threaded = true;
+}; // struct FrustumRasterizerSettings
+
+
+struct FrustumRasterizerMask
+{
+    using AccessorT = const tree::ValueAccessor<const MaskTree>;
+
+    FrustumRasterizerMask() = default;
+
+    explicit FrustumRasterizerMask(
+        const math::Transform& transform,
+        const MaskGrid* mask = nullptr,
+        const BBoxd& bbox = BBoxd(),
+        const bool clipToFrustum = true,
+        const bool invert = false);
+
+    operator bool() const;
+
+    MaskTree::ConstPtr getTreePtr() const;
+
+    bool valid(const Coord& ijk, AccessorT* acc) const;
+
+    const CoordBBox& clipBBox() const;
+
+private:
+    MaskGrid::Ptr mMask;
```

Comment:
make unique ptr? This object owns this grid and can return const raw ptrs to avoid shared atomics. You'll have to std::move it in `FrustumRasterizer`.

Reply:
Ah I see, I didn't realise the second use case - yeah I think even if you could work around it that using shared_ptr is the correct tool here!